### DISCUSSION TITLE
Dashboarding enhancements

### DISCUSSION THREAD

**Discussion by nblumhardt** [datalust/seq-tickets#1111] [Ideas]
Tracking issue for a number of things that didn't make the 4.0 cut.

- [x] Time slice duration selector: instead of just providing _Last 30 days_, also permit _by 1 day_, _by 1 hr_ etc. to be selected. #581
- [x] Double-height and double-width charts. #584
- [x] Make the SQL shown on a chart's _SQL_ tab clickable, to open the query directly in the _events_ screen. #570 
- [x] 15 and 30 minute refresh interval options. #580
- [x] Additional color palettes. #587
- [x] Don't refresh the dashboard when not visible. #588
- [x] Single-valued charts, i.e. show one value computed over the whole dashboard range - this might not necessarily be numeric, so that things like circuit breaker status (text) can be shown. #591
- [x] Show a nice subtle icon instead of `null` when no data is present in a time range. #627
- [x] Thresholds - show dotted line on chart, or similar. #630
- [x] Pie charts - also computed over the whole range. #629
- [x] Control legend position - hide, move, float... #730
- [x] Hide items from the dashboard list, in the same way that only a subset of signals are shown in the Signal Bar. #626
- [ ] ~Make chart titles easier to read at low resolutions, e.g. wall-mounted TVs.~ #625
- [x] Rowset-valued charts, i.e. show a tabular result computed over the whole dashboard range. ~#628~
- [x] Login-free dashboards; e.g. put an API key parameter in the dashboard URL so that it can be viewed on information radiators without needing a logged-in Seq session. #799
- [x] `having`, `order by` and `limit` support on charts #798
- [ ] Cycle through several dashboards; also for information radiators, allow multiple dashboards to be selected and cycle through them at whatever refresh interval is specified.
- [ ] Copy or move a chart to a different dashboard
- [ ] Multiple scales per chart
- [ ] Annotations - add notes to a point on a chart
- [ ] _by second_ interval option for 5/15-minute time range #1255
- [ ] Drag-and-drop widget arrangement
- [ ] Overlays of arbitrary single-value aggregations - `min`, `max` etc.
- [ ] Request a chart as .PNG through the Seq API
- [ ] Original creator should have rights to un-share or edit a dashboard after it is shared
- [ ] Admin tooling to change dashboard ownership
- [ ] Stacked bar charts
- [ ] Custom dashboard time ranges, e.g. a specific range of days, weeks, months, years at some time in the past #1255
- [ ] Charts with logarithmic scales
- [ ] Text/markdown "charts"
- [ ] Custom dashboard refresh intervals
- [ ] "Traffic light" charts, optimized for health check reporting
- [ ] On numeric, single-value charts, show a small up/down or +/- N trend indicator (delta since last refresh)
- [ ] Units on values: kWh, $, etc.
- [ ] Duplicate an alert
- [ ] Logarithmic scales for y axes
- [ ] Independent axes for different series
- [ ] Option to show values or percentages in segments of pie charts
- [ ] Render the current bucket differently in timeseries charts to indicate that data may be incomplete
 
I'm sure there's more to add here, we'll update this list as feedback comes in :-)

**Comment by nblumhardt** [comment]
Folks tracking this issue, if you have a favorite in the list above, could you please make a mention in the comments? Some may carry over to 4.2 and beyond, so feedback will help prioritize these. Thanks :-)

**Comment by andymac4182** [comment]
Additional color palettes for warnings would be great

**Comment by nblumhardt** [comment]
4.1 is getting substantial - there's quite a bit from this ticket and others that we'd like to start getting into people's hands: https://github.com/datalust/seq-releases/milestone/75?closed=1.

I've broken out one more of the TODO items from above - single-valued charts - and assigned this to 4.1. I think the rest of this ticket can move to 4.2. Please jump in if you're hoping to see any of the incomplete items above in the 4.1 milestone. Thanks! :-)

**Comment by andymac4182** [comment]
It would be great if you could have two scales on a single chart. I have the min, max, mean of query times and the count of the number of queries. The aggregates are in the 1-5 mark while the count is up at 100-200 mark.

**Comment by tsimbalar** [comment]
What I would like to have is some kind of comparing a chart (or a complete dashboard) with the same one at a different period in the past. 
This would help to determine if the behavior is consistent with what happened in the past.

Typical cases are : compare this week with last week or the week before last / compare this month with last month etc .

I can't really think exactly of how that would materialize in the UI though :-/ 
Possible options , but not really sure : 
- inside a chart with a result of query Q on a given timerange T, also render the result of Q on timerange T-1 (if T is last 7 days, T-1 is from D-14 to D-7 etc)
- add at the chart configuration level : an optional "Compare with another time range" that would add another chart just under the original one, but with a different time period .... and maybe options like "previous time range" (automatically deduce the period to compare with, depending on the size of the timerange selected in the dashboard's picker ) , "time range of same duration starting at *this hard-coded date*" , "time range of same duration starting at the *dashboard start date minus X days/minutes/seconds" ...
-  at the dashboard level : "show side by side with this dashboard at period "N-1""

... I like option 2 best :) 

I guess that for this to make it really useful , Seq would need to also support timereanges other than "last X days" and gain some kind of understanding about "when a week starts" (the good old "Sunday" or "Monday" -based :p) , "when did the month start", in order to have time ranges like "this week", "this month", "this year", "last month", "last week", "the week before last" etc ...

**Comment by stanisls** [comment]
I'd vote for making Dashboard charts available as rendered images (PNG or, maybe, SVG) through API. This would be a great way to present log-based status in third-party monitoring applications, as well as providing visual access to application metric to departments, that are not necessarily involved in development.

**Comment by nskerl** [comment]
It is somewhat annoying that the selected (or deselected) properties on a chart are reset when the Refresh fires.  I expect that they would remain in the prior state between refreshes. Maybe this is something we can add to the dashboard enhancements.

Thanks

![seq-reset](https://user-images.githubusercontent.com/782187/33637130-1af18298-d9d4-11e7-98e4-86f6dc6b6b13.gif)

**Comment by janpieterz** [comment]
Super interested in these two:

 Login-free dashboards; e.g. put an API key parameter in the dashboard URL so that it can be viewed on information radiators without needing a logged-in Seq session.
Cycle through several dashboards; also for information radiators, allow multiple dashboards to be selected and cycle through them at whatever refresh interval is specified.

**Comment by whoiskevinrich** [comment]
I'd like the ability to "lock" a dashboard's range & group by dashboard ID. There are some signals that only get fired daily and some that are fired on the minute, but if one dashboard is set to "Last 2 days by hour", it seems all dashboards are.

**Comment by jonnepmyra** [comment]
Hi, any updates to the "Request a chart as .PNG through the Seq API" feature? We'd really love to show our end users some of our dashboard widgets containing important metrics from our backend systems in their frontend apps. Really looking forward to this feature!

**Comment by nblumhardt** [comment]
Hi @jonnepmyra - thanks for the nudge. No movement on this one as yet, but it's definitely on the radar :+1:

**Comment by stanisls** [comment]
Hi. Wondering if any progress is made on "Request a chart as .PNG through the Seq API" feature request?

**Comment by nblumhardt** [comment]
@stanisls not currently, but thank you for the +1 👍

**Comment by andymac4182** [comment]
It would be great if you could allow the legend to go larger. We have some dashboards where there is a large amount of room spare around a pie chart that could allow the legend to fill the space and show all the labels.

**Comment by Sebbl22** [comment]
I would like to vote for the custom time ranges, since we like to realize some kind of reporting, e.g. "How many errors | warnings | log messages | \<custom properties\> did we have today | yesterday | last week".
It is easily doable with querys, but much more time-consuming.

> Custom dashboard time ranges, e.g. a specific range of days, weeks, months, years at some time in the past

**Comment by andymac4182** [comment]
Another thought I had was it would be nice to be able to have 2 time ranges for comparison. Eg Today vs Yesterday.

**Comment by bard83** [comment]
Nice to have: refresh single chart in a dashboard, when it goes in timeout.
![dashboard](https://user-images.githubusercontent.com/11898472/80470631-92164000-8942-11ea-9aa3-deac918b8426.png)

**Comment by nblumhardt** [comment]
@bard83 thanks for the suggestion!

If you're hitting timeouts when rendering a chart, sometimes moving the condition from _Query > `where`_ into a signal, and using the signal to select data for the chart, will speed things up (signals are indexed, while general `where` clauses won't use indexing).

If that doesn't help, please feel free to get in touch with `support@` and send _Settings_ > _Diagnostics_ > _Download a full diagnostic report_. Sometimes there might be a hardware bottleneck, but in other cases we can spot simple improvements that can help.

**Comment by tbolon** [comment]
- [ ] Ability to embed a dashboard without any setting or decoration (no title, no time / refresh selection, only graphs)
- [ ] Ability to embed a single graph from a dashboard (again no title, no time selection, only the panel)

We want to embed some dashboard in our supervision website, but the current maximized view still use a lot of space.

**Comment by PassivePicasso** [comment]
It would be nice to be able to have charts in a Dashboard have the ability to interact with the Dashboards time settings differently.
I want to create charts for things like "Occurrences of opened file last month organized by count descending" for showing things like popular usages of features/content, etc.
